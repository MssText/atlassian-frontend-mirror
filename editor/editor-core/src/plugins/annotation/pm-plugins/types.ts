import { PortalProviderAPI } from '../../../ui/PortalProvider';
import { Dispatch, EventDispatcher } from '../../../event-dispatcher';
import { AnnotationTypeProviders } from '../types';

export interface InlineCommentPluginOptions {
  dispatch: Dispatch;
  eventDispatcher: EventDispatcher;
  portalProviderAPI: PortalProviderAPI;
  inlineCommentProvider: AnnotationTypeProviders['inlineComment'];
  pollingInterval?: number;
}
