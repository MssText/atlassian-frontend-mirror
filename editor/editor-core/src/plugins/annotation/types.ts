import React from 'react';
import { AnnotationType } from '@atlaskit/adf-schema';
import { ACTIONS } from './pm-plugins/actions';
import { DecorationSet } from 'prosemirror-view';
import { EditorState, SelectionBookmark } from 'prosemirror-state';

export type AnnotationInfo = {
  id: string;
  type: AnnotationType;
};

export enum AnnotationTypes {
  INLINE_COMMENT = 'inlineComment',
}

type AnnotationComponentProps = {
  /**
   * Selected text (can be used when creating a comment)
   */
  textSelection?: string;

  /**
   * DOM element around selected text (for positioning)
   */
  dom?: HTMLElement;
};

export type AnnotationCreateComponentProps = AnnotationComponentProps & {
  /**
   * Creates an annotation mark in the document with the given id.
   */
  onCreate: (id: string) => void;

  /**
   * Indicates that a draft comment was discarded/cancelled
   */
  onClose?: () => void;
};

export type AnnotationViewComponentProps = AnnotationComponentProps & {
  /**
   * Existing annotations where the cursor is placed.
   * These are provided in order, inner-most first.
   */
  annotations: Array<AnnotationInfo>;

  /**
   * Resolves an annotation with the given ID around the selection.
   */
  onResolve: (id: string) => void;

  /**
   * Removes the annotation from the document
   */
  onDelete?: (id: string) => void;
};

export interface AnnotationState<Type, State> {
  annotationType: Type;
  id: string;
  state: State;
}

export interface AnnotationTypeProvider<Type, State> {
  getState: (
    annotationIds: string[],
  ) => Promise<AnnotationState<Type, State>[]>;
  pollingInterval?: number;
}

export type InlineCommentState = { resolved: boolean };
export type AnnotationTypeProviders = {
  inlineComment: AnnotationTypeProvider<
    AnnotationTypes.INLINE_COMMENT,
    InlineCommentState
  >;
};

export interface AnnotationProvider {
  providers?: AnnotationTypeProviders;
  createComponent?: React.ComponentType<AnnotationCreateComponentProps>;
  viewComponent?: React.ComponentType<AnnotationViewComponentProps>;
}

export type InlineCommentMap = { [key: string]: boolean };
export type InlineCommentAction =
  | {
      type: ACTIONS.INLINE_COMMENT_RESOLVE;
      data: { id: string };
    }
  | {
      type: ACTIONS.SET_INLINE_COMMENT_STATE;
      data: InlineCommentMap;
    }
  | {
      type: ACTIONS.SET_INLINE_COMMENT_DRAFT_STATE;
      data: {
        drafting: boolean;
        editorState: EditorState;
      };
    }
  | {
      type: ACTIONS.INLINE_COMMENT_UPDATE_MOUSE_STATE;
      data: {
        mouseData: {
          x?: number;
          y?: number;
          isSelecting?: boolean;
        };
      };
    };

export type InlineCommentPluginState = {
  annotations: InlineCommentMap;
  mouseData: {
    x: number;
    y: number;
    isSelecting: boolean;
  };
  draftDecorationSet?: DecorationSet;
  bookmark?: SelectionBookmark<any>;
};

const prefix = 'ak-editor-annotation';
export const AnnotationTestIds = {
  prefix,
  floatingComponent: `${prefix}-floating-component`,
  floatingToolbarCreateButton: `${prefix}-toolbar-create-button`,
};

export const DraftDecorationClassName = `${prefix}-draft-selection`;
