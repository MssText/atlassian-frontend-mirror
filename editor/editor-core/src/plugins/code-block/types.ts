export interface CodeBlockOptions {
  allowCopyToClipboard?: boolean;
}
