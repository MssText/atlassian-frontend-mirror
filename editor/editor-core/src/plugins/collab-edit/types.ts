import { ReactElement } from 'react';
import { Providers } from '@atlaskit/editor-common/provider-factory';
import { CollabEditProvider } from '@atlaskit/editor-common';

export {
  CollabParticipant,
  CollabEventInitData,
  CollabEventRemoteData,
  CollabEventConnectionData,
  CollabeEventPresenceData,
  CollabEventTelepointerData,
  CollabSendableSelection,
} from '@atlaskit/editor-common';

export type InviteToEditComponentProps = {
  children: ReactElement<InviteToEditButtonProps>;
};

export type InviteToEditButtonProps = {
  onClick: (event: React.MouseEvent<HTMLElement>) => void;
  selected: boolean;
};

export interface CollabInviteToEditProps {
  inviteToEditHandler?: (event: React.MouseEvent<HTMLElement>) => void;
  isInviteToEditButtonSelected?: boolean;
  inviteToEditComponent?: React.ComponentType<InviteToEditComponentProps>;
}

export type CollabEditOptions = {
  provider?: Providers['collabEditProvider'];
  userId?: string;
  useNativePlugin?: boolean;
  allowUnsupportedContent?: boolean;
  sendDataOnViewUpdated?: boolean;
} & CollabInviteToEditProps;

export type PrivateCollabEditOptions = CollabEditOptions & {
  sanitizePrivateContent?: boolean;
};

export type ProviderCallback = (
  codeToExecute: (provider: CollabEditProvider) => void,
) => void;

export type ProviderBuilder = (
  collabEditProviderPromise: Promise<CollabEditProvider>,
) => ProviderCallback;
