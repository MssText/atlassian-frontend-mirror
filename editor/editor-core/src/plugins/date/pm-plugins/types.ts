export type DatePluginState = {
  showDatePickerAt: number | null;
};
