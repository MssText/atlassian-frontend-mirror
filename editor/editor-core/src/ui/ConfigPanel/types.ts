import {
  ExtensionProvider,
  ExtensionType,
  ExtensionKey,
  Parameters,
} from '@atlaskit/editor-common/extensions';

export type FormResult = {
  [key: string]: string | number | string[] | number[] | undefined;
};

export enum ValidationError {
  Required = 'required',
}

export enum FieldTypeError {
  isMultipleAndRadio = 'isMultipleAndRadio',
}

export type Entry<T> = [string, T];

export type PublicProps = {
  extensionProvider: ExtensionProvider;
  extensionType: ExtensionType;
  extensionKey: ExtensionKey;
  nodeKey: string;
  extensionParameters?: Parameters;
  parameters?: Parameters;
  autoSave?: boolean;
  closeOnEsc?: boolean;
  showHeader?: boolean;
  onChange: (data: Parameters) => void;
  onCancel: () => void;
};

export type OnBlur = (name: string) => void;
