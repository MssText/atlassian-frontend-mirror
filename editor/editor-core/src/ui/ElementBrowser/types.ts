export type Category = {
  title: string;
  name: string;
};

export enum Modes {
  full = 'full',
  inline = 'inline',
}
