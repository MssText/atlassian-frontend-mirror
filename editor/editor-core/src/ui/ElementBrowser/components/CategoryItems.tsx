import React, { memo, useCallback } from 'react';
import styled, { css, ThemeProvider } from 'styled-components';
import { QuickInsertItem } from '@atlaskit/editor-common/provider-factory';
import { colors } from '@atlaskit/theme';
import Item from '@atlaskit/item';
import { withAnalyticsContext } from '@atlaskit/analytics-next';
import { Shortcut } from '../../styles';
import { ItemIcon } from '../../../plugins/type-ahead/ui/TypeAheadItemsList';
import IconFallback from '../../../plugins/quick-insert/assets/fallback';
import {
  GRID_SIZE,
  SCROLLBAR_THUMB_COLOR,
  SCROLLBAR_TRACK_COLOR,
  FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS,
} from '../constants';
import { Modes } from '../types';
import useContainerWidth from '../hooks/useContainerWidth';

interface Props {
  items: QuickInsertItem[];
  mode: keyof typeof Modes;
  onSelectItem: (item: QuickInsertItem) => void;
  onEnterKeyPress: (item: QuickInsertItem) => void;
}

function CategoryItems({ items, mode, ...props }: Props) {
  const { containerWidth, ContainerWidthMonitor } = useContainerWidth();

  const itemTheme = {
    ...(mode === Modes.full && {
      width: {
        default: getItemWidthForContainer(containerWidth),
      },
      '-ms-flex': 'auto',
      position: 'relative',
      boxSizing: 'border-box',
    }),
    height: {
      default: GRID_SIZE * GRID_SIZE,
    },
    padding: {
      default: {
        top: GRID_SIZE * 1.5,
        right: GRID_SIZE * 1.5,
        bottom: GRID_SIZE * 1.5,
        left: GRID_SIZE * 1.5,
      },
    },
    borderRadius: GRID_SIZE / 2,
    hover: {
      background: 'rgb(244, 245, 247)',
    },
    beforeItemSpacing: {
      default: GRID_SIZE * 1.5,
    },
  };
  return (
    <>
      <ContainerWidthMonitor />
      <CategoryItemsWrapper mode={mode}>
        <ThemeProvider theme={{ '@atlaskit-shared-theme/item': itemTheme }}>
          <>
            {items.map<JSX.Element>(
              (item, index): JSX.Element => (
                <CategoryItem key={index} item={item} {...props} />
              ),
            )}
          </>
        </ThemeProvider>
      </CategoryItemsWrapper>
    </>
  );
}

const getItemWidthForContainer = (clientWidth?: number) => {
  if (clientWidth == null) return '33%';
  if (clientWidth < FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.small)
    return '100%';
  if (
    clientWidth >= FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.small &&
    clientWidth < FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.medium
  ) {
    return '50%';
  }
  if (
    clientWidth >= FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.medium &&
    clientWidth < FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.large
  ) {
    return '33%';
  }
  if (clientWidth >= FLEX_ITEMS_CONTAINER_BREAKPOINT_NUMBERS.large)
    return '25%';
};

type CategoryItemType = {
  item: QuickInsertItem;
  onSelectItem: (item: QuickInsertItem) => void;
  onEnterKeyPress: (item: QuickInsertItem) => void;
};

function CategoryItem({
  item,
  onSelectItem,
  onEnterKeyPress,
}: CategoryItemType) {
  const onClick = useCallback(() => onSelectItem(item), [item, onSelectItem]);
  const onKeyPress = useCallback(
    (e: KeyboardEvent) => {
      if (e.key !== 'Enter') return;
      onEnterKeyPress(item);
    },
    [item, onEnterKeyPress],
  );
  return (
    <Item
      onClick={onClick}
      onKeyPress={onKeyPress}
      elemBefore={
        <ItemIcon>
          {item.icon ? item.icon() : <IconFallback label={item.title} />}
        </ItemIcon>
      }
      isSelected={item.selected}
      aria-describedby={item.title}
    >
      <ItemBody>
        <ItemText>
          <div>{item.title}</div>
          {item.description && (
            <ItemDescription>{item.description}</ItemDescription>
          )}
        </ItemText>
        <ItemAfter>
          {item.keyshortcut && <Shortcut>{item.keyshortcut}</Shortcut>}
        </ItemAfter>
      </ItemBody>
    </Item>
  );
}

interface ContainerProps {
  mode: keyof typeof Modes;
}

const scrollbarStyle = css`
  ::-webkit-scrollbar {
    width: ${GRID_SIZE}px;
  }
  ::-webkit-scrollbar-track-piece {
    background: ${SCROLLBAR_TRACK_COLOR};
  }
  ::-webkit-scrollbar-thumb {
    background: ${SCROLLBAR_THUMB_COLOR};
  }

  /** Firefox **/
  scrollbar-color: ${SCROLLBAR_THUMB_COLOR} ${SCROLLBAR_TRACK_COLOR};

  -ms-overflow-style: -ms-autohiding-scrollbar;
`;

const CategoryItemsWrapper = styled.div`
  display: ${({ mode }: ContainerProps) =>
    mode === Modes.full ? 'flex' : 'block'};
  flex-flow: row wrap;
  align-items: flex-start;
  justify-content: flex-start;
  overflow-x: hidden;
  overflow-y: scroll;
  ${scrollbarStyle};
`;

const ItemBody = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  justify-content: space-between;
  line-height: 1.4;
  width: 100%;
`;

const ItemDescription = styled.div`
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  font-size: 11.67px;
  color: ${colors.N200};
  margin-top: ${GRID_SIZE / 2}px;
`;

const ItemText = styled.div`
  width: inherit;
  white-space: initial;
`;

const ItemAfter = styled.div`
  flex: 0 0 auto;
`;

const MemoizedCategoryItemsWithAnalytics = memo(
  withAnalyticsContext({ component: 'category-item' })(CategoryItems),
);

export default MemoizedCategoryItemsWithAnalytics;
