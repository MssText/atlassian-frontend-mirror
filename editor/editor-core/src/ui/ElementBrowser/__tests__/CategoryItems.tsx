const getWidth = jest.fn();

jest.mock('../hooks/useContainerWidth', () => () => ({
  containerWidth: getWidth(),
  ContainerWidthMonitor: jest.fn(),
}));

import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import CategoryItems from '../components/CategoryItems';
import { Modes } from '../types';

const props = {
  items: [
    { name: 'item-1', title: 'Item 1', action: jest.fn() },
    { name: 'item-2', title: 'Item 2', action: jest.fn() },
    { name: 'item-3', title: 'Item 3', action: jest.fn() },
    { name: 'item-4', title: 'Item 4', action: jest.fn() },
    { name: 'item-5', title: 'Item 5', action: jest.fn() },
  ],
  mode: Modes.full,
  onSelectItem: jest.fn(),
  onEnterKeyPress: jest.fn(),
};

const shallowRenderWithCustomWidth = (width: number): ShallowWrapper => {
  getWidth.mockReturnValueOnce(width);
  // 3 dives to skip analytics, context provider, and to get ThemeProvider
  const wrapper = shallow(<CategoryItems {...props} />);
  return wrapper
    .dive()
    .dive()
    .dive();
};

const getThemeProviderWidth = (width: number) => {
  const res: any = shallowRenderWithCustomWidth(width);
  return res.find('ThemeProvider').prop('theme')['@atlaskit-shared-theme/item']
    .width.default;
};

describe('CategoryItems', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(<CategoryItems {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  describe('getItemWidthForContainer', () => {
    it('should get 50% width for an item if the container width is >= 400px', () => {
      expect(getThemeProviderWidth(500)).toStrictEqual('50%');
    });
    it('should get 33% width for an item if the container width is >= 600px and < 1024px', () => {
      expect(getThemeProviderWidth(601)).toStrictEqual('33%');
    });
    it('should get 25% width for an item if the container width is >= 1024px', () => {
      expect(getThemeProviderWidth(1025)).toStrictEqual('25%');
    });
  });
  it('should render CategoryItem component', () => {
    const wrapper = shallow(<CategoryItems {...props} />);
    const itemWrapper = wrapper
      .dive()
      .dive()
      .dive()
      .find('CategoryItem')
      .first();
    expect(itemWrapper).toMatchSnapshot();
  });
});
