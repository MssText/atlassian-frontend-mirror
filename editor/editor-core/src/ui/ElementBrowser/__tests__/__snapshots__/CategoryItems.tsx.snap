// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryItems should match snapshot 1`] = `
<AnalyticsContext
  data={
    Object {
      "component": "category-item",
    }
  }
>
  <CategoryItems
    items={
      Array [
        Object {
          "action": [MockFunction],
          "name": "item-1",
          "title": "Item 1",
        },
        Object {
          "action": [MockFunction],
          "name": "item-2",
          "title": "Item 2",
        },
        Object {
          "action": [MockFunction],
          "name": "item-3",
          "title": "Item 3",
        },
        Object {
          "action": [MockFunction],
          "name": "item-4",
          "title": "Item 4",
        },
        Object {
          "action": [MockFunction],
          "name": "item-5",
          "title": "Item 5",
        },
      ]
    }
    mode="full"
    onEnterKeyPress={[MockFunction]}
    onSelectItem={[MockFunction]}
  />
</AnalyticsContext>
`;

exports[`CategoryItems should render CategoryItem component 1`] = `
<CategoryItem
  item={
    Object {
      "action": [MockFunction],
      "name": "item-1",
      "title": "Item 1",
    }
  }
  key="0"
  onEnterKeyPress={[MockFunction]}
  onSelectItem={[MockFunction]}
/>
`;
